### PR TITLE
#162949079 User can like and dislike an article

### PR DESCRIPTION
**Pull request by jomasim**
### What does this PR do?
Allows a user to like and dislike an article.

### Description of Task to be completed?
- [x] Implement liking an article
- [x] Implement  disliking an article
- [x] Add like and dislike article tests

### How should this be manually tested?
1. Clone this repository on your local machine using 

     `git clone https://github.com/andela/ah-robotics.git`
2. Change your working directory to the project's directory using 
     
     `cd ah-robotics`
3. Change to the specific branch `git checkout ft-article-like-dislike-162949079` using 

     `git checkout ft-article-like-dislike-162949079 `
4. Create a virtual environment using the following command: 
     
     `python3 -m virtualenv env`
5. Activate the virtual environment using 
     
     `source env/bin/activate`
6. Install all the requirements using 
     
     `pip install -r requirements.txt`

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
        `pg_ctl -D /usr/local/var/postgres start`
    
    2. Run psql interactive terminal
    
        `psql postgres`
    
    3. Create a new user:
    
        `CREATE USER sample_username WITH PASSWORD 'sample_password';`
    
    4. Grant privileges to the user:
    
        `ALTER USER sample_username CREATEDB;`
    
    5. Create a database:
    
        `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
    
    6. Exit psql interactive with the following command: `\q`
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      `source .env`
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      `python manage.py makemigrations`
      
      (ii) Apply the migrations using: 
      
      `python manage.py migrate`

10. Run the following command to collect static files.

    `python manage.py collectstatic`
      
#### Testing the feature
1. Run the server from the terminal using: 
    
    `python manage.py runserver`
2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and the add the following payload to the body:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint 
```
http://127.0.0.1:8000/api/v1/users/login/
```
Add the following payload to the body section.
    
```
        {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
 ```
7. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
8. After logging in the screen should look as follows:
    
    
screenshot 2019-01-29 at 16 26 20

9. Copy the token and click on the **Headers** section. In the **Key** column, type in the word **Authorization**. And in its corresponding value in the **Value** column, write the word __token__, add a space and paste in the token that you copied from the login response. The **Headers** section should look as follows after doing so:
    
    image

10. To Create an article, insert the following URL: 
```
http://127.0.0.1:8000/api/v1/articles/ 
```
In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
```
    {
        "article": {
            "title": "Sample Article",
            "description": "Sample Article Description",
            "body": "The sample article body"
        }
    }
```
 
Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.

11. To like an article, insert the following URL:
 ```
 http://127.0.0.1:8000/api/v1/articles/<slug>/like/
```
 ```
  e.g. http://127.0.0.1:8000/api/v1/articles/sample-article/like/
```
 In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, click **Send** to post a like.
12 . To like an article, insert the following URL:
 ```
 http://127.0.0.1:8000/api/v1/articles/<slug>/dislike/
```
 ```
  e.g. http://127.0.0.1:8000/api/v1/articles/sample-article/dislike/
```
 In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, click **Send** to post a dislike.

#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#162949079](https://www.pivotaltracker.com/n/projects/2235171/stories/162949079)

### Screenshots
sample response for posting a like
screenshot 2019-01-29 at 17 56 54
sample response for posting a dislike 
screenshot 2019-01-29 at 18 06 16
## Questions:
Questions and comments are highly welcome.

### CODE REVIEW

**Comment by Teatoller** [comment]
Hello Joma,

Your pull request is succinct.
I have been able to follow through step-by-step and tested the application successfully.
Good work

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @jomasim,
I have reviewed the work and the feature works well as expected. Good work.

**Review by waracci — COMMENTED** [review]
Works as expected. 👍

**Review by johnwayodi — COMMENTED** [review]
> 12 . To like an article, insert the following URL:
 >```
 >http://127.0.0.1:8000/api/v1/articles/<slug>/dislike/
>```
>```
>  e.g. http://127.0.0.1:8000/api/v1/articles/sample-article/dislike/
>```

Tested and works well,
although you need to make a minor correction on instruction 12 on the PR, thanks.

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -5,6 +5,52 @@
...
+
+
+class ReactionManager(models.Manager):
+    """
+    The class enables article likes and dislikes set to be managed separately.
+    """
+    use_related_fields = True
+
+    def likes(self):
+        "fetches reactions with a value greater than zero"
```
Comment (sharkdevs) [inline authors/apps/articles/models.py]: Ensure your docstring comments have three pairs of double quotation marks in this file.

File: authors/apps/articles/models.py [hunk 2]
```diff
@@ -5,6 +5,52 @@
...
+        return self.get_queryset().filter(reaction__gt=0)
+
+    def dislikes(self):
+        "fetches reactions with a value less than zero"
```
Comment (sharkdevs) [inline authors/apps/articles/models.py]: docstring

File: authors/apps/articles/models.py [hunk 3]
```diff
@@ -5,6 +5,52 @@
...
+        return self.get_queryset().filter(reaction__lt=0)
+
+    def sum_rating(self):
+        "returns the sum of reaction items"
```
Comment (sharkdevs) [inline authors/apps/articles/models.py]: docstring

File: authors/apps/articles/urls.py [hunk 4]
```diff
@@ -1,11 +1,18 @@
...
+from .models import Article, Reaction
 
 app_name = "articles"
 
 urlpatterns = [
     path('articles/', views.ListCreateArticle.as_view(), name='articles'),
     path('articles/<slug>/', views.RetrieveUpdateDeleteArticle.as_view(),
          name='article-details'),
+    path('articles/<slug>/like/', views.ReactionView.as_view(
+        model=Article, reaction_type=Reaction.LIKE),
```
Comment (sharkdevs) [inline authors/apps/articles/urls.py:13]: Unlike our other urls, you have models and reaction_type fields added to your urls.  What does this do? (especially the model part)

Comment (jomasim) [reply authors/apps/articles/urls.py:13]: reaction type is passed directly on the view when a request is sent to the endpoint. This defines the action type whether it is a  like or dislike.

**Comment by Collins33** [comment]
Good work on the task. Works as expected

**Review by Collins33 — APPROVED** [review]

**Pull request merged**